app/controllers/webauthn_credentials_controller.rb

```diff
@@ -0,0 +1,57 @@
+class WebauthnCredentialsController < ApplicationController
+  before_action :redirect_to_signin, unless: :signed_in?
+
+  def create
+    @create_options = current_user.webauthn_options_for_create
+
+    session[:webauthn_registration] = { "challenge" => @create_options.challenge }
+
+    render json: @create_options
+  end
+
+  def callback
+    webauthn_credential = build_webauthn_credential
+
+    if webauthn_credential.save
+      flash[:notice] = t(".webauthn_credential.saved")
+      redirect_to edit_settings_path
+    else
+      message = webauthn_credential.errors.full_messages.to_sentence
+      render json: { message: message }, status: :unprocessable_entity
+    end
+  rescue WebAuthn::Error => e
+    render json: { message: e.message }, status: :unprocessable_entity
+  ensure
+    session.delete("webauthn_registration")
+  end
+
+  def destroy
+    webauthn_credential = current_user.webauthn_credentials.find(params[:id])
+    if webauthn_credential.destroy
+      flash[:notice] = t(".webauthn_credential.confirm_delete")
+    else
+      flash[:error] = webauthn_credential.errors.full_messages.to_sentence
```

Comment:
These were persisting because in a[ previous commit](https://github.com/rubygems/rubygems.org/pull/2865/commits/4155a02e60eca330b290cc7f4a25606cf2e07119#diff-9758e4d5f8fcd6638b957895502b8058bd93fe317536d6adc517de90e8ea964cR8) I added a `flash.keep` to edit settings in an attempt to get the [creation flash](https://github.com/rubygems/rubygems.org/pull/2865/commits/4155a02e60eca330b290cc7f4a25606cf2e07119#diff-31ed563a2312260475e634c0d80241ff4f764d163e7612f5e1c7b6be332c6d91R16) to stick around. 

When creating a security device, we call redirect [here](https://github.com/rubygems/rubygems.org/pull/2865/files#diff-31ed563a2312260475e634c0d80241ff4f764d163e7612f5e1c7b6be332c6d91R16) in the callback, but we don't actually redirect directly. Rather, we send that as a response to the javascript callback method. The JS then [handles that response](https://github.com/rubygems/rubygems.org/pull/2865/files#diff-31ed563a2312260475e634c0d80241ff4f764d163e7612f5e1c7b6be332c6d91R16) by [setting our location](https://github.com/rubygems/rubygems.org/pull/2865/files#diff-b0f7b986fa0b7edf061b91cf0be6d804f4f2607e73043e8ee35bc97a1b0f81ceR15) to the desired redirect url. However, because of this we lose the flash message. Adding flash.keep was an attempt to get that flash message to stick around, but as you noticed, it causes all of the messages on this page to persist through too many redirects and so wasn't the right solution.

I looked at this extensively and paired on it with @bettymakes . We could find no way that wasn't extremely hacky to persist a flash message through the js callback on security device creation. After spending a lot of time on this, we decided that the impact of shipping with the flash is low compared to the additional time/effort we’ll need to sink into this. So, for now it is better to just [not have](https://github.com/rubygems/rubygems.org/pull/2865/commits/abba1831b73fe4a96f0272436132cb176bc62a8d) a flash on create. I've removed the `flash.keep` that caused the deletion message to stick around and we can save adding a flash on create for future exploration.